pkg/controller/nodedaemon/daemonsets.go

```diff
@@ -0,0 +1,254 @@
+package nodedaemon
+
+import (
+	"context"
+
+	"github.com/openshift/local-storage-operator/pkg/controller/util"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// CreateOrUpdateDaemonset fetches and creates/updates ds according to desiredDs
+// fields that are not allowed update at runtime are not copied unless the creationtimestamp is zero
+func CreateOrUpdateDaemonset(c client.Client, desiredDs *appsv1.DaemonSet) (controllerutil.OperationResult, error) {
+	ds := &appsv1.DaemonSet{ObjectMeta: metav1.ObjectMeta{Name: desiredDs.GetName(), Namespace: desiredDs.GetNamespace()}}
+	return controllerutil.CreateOrUpdate(context.TODO(), c, ds, func() error {
+		updateDaemonSet(*desiredDs, ds)
+		return nil
+	})
+}
+
+// updateDaemonSet copies the daemonset from in to out
+// fields that are not allowed update at runtime are not copied unless the creationtimestamp is zero
+func updateDaemonSet(in appsv1.DaemonSet, out *appsv1.DaemonSet) {
+
+	if out.CreationTimestamp.IsZero() {
+		in.DeepCopyInto(out)
+		return
+	}
+	out.ObjectMeta.Labels = in.GetLabels()
+	out.ObjectMeta.Annotations = in.ObjectMeta.Annotations
+	out.ObjectMeta.OwnerReferences = in.ObjectMeta.OwnerReferences
+
+	out.Spec.Template.ObjectMeta.Annotations = in.Spec.Template.ObjectMeta.Annotations
+	out.Spec.Template.Spec = in.Spec.Template.Spec
+
+	out.Spec.UpdateStrategy = in.Spec.UpdateStrategy
+	out.Spec.MinReadySeconds = in.Spec.MinReadySeconds
+	out.Spec.RevisionHistoryLimit = in.Spec.RevisionHistoryLimit
+	return
```

Comment:
I checked the way you apply DaemonSet and I can immediately spot an issue: what if user (or another operator / controller / whatever) changes Spec.Selector? This code does not fix the selector back to the state the operator wants it. This may result in unexpected and **very** hard to debug behavior of the whole LSO.

Not mentioning that you're abusing the API server by calling Update even when there may be no actual update necessary.

Of course, you can add the selector to things the operator replaces. But explicitly listing fields you're going to fix is very brittle. In the future, Kubernetes may introduce a new DaemonSet.Spec field that the operator does not know about and it will not fix it. And it may break the operator again.

What is 100% safe and future proof is to remember the last Generation of the DaemonSet that LSO created / updated. When the generation of the object changes, the operator can compare the current Generation with the last correct Generation and immediately see that someone / something has changes DaemonSet.Spec and overwrite the whole spec. And that's why we're tracking generations of children in LSO Status.

Reply:
Copied this commonet over to #110 and replied there: 
https://github.com/openshift/local-storage-operator/pull/110#discussion_r457153757